File: src/test/java/com/ddubson/repositories/BookHQLRepositoryTest.java
package com.ddubson.repositories;

import com.ddubson.models.Book;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.test.annotation.DirtiesContext;
import org.springframework.test.context.junit4.SpringRunner;

import java.util.List;

import static java.util.Arrays.asList;
import static java.util.stream.Collectors.toList;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.junit.Assert.assertThat;
import static org.springframework.test.annotation.DirtiesContext.ClassMode.BEFORE_CLASS;

@RunWith(SpringRunner.class)
@DataJpaTest
@AutoConfigureTestDatabase
@DirtiesContext(classMode = BEFORE_CLASS)
public class BookHQLRepositoryTest {
    @Autowired
    BookHQLRepository bookRepository;

    @Test
    public void findByTitle_shouldReturnBookWithSpecifiedTitle() {
        String expectedTitle = "War and Peace";
        Book book = bookRepository.findByTitle(expectedTitle);
        assertThat(book.getTitle(), equalTo(expectedTitle));
    }

    @Test
    public void findByAuthorId_returnsAllBooksOfAuthor() {
        List<String> expectedBooks = asList("Anna Karenina", "Crime and Punishment");
        List<Book> books = bookRepository.findByAuthorId(4L);
        assertThat(books.stream().map(Book::getTitle).collect(toList()),
                equalTo(expectedBooks));
    }

    @Test
    public void findByAuthorLastName_returnsAllBooksOfAuthor() {
        List<String> expectedBooks = asList("Anna Karenina", "Crime and Punishment");
        List<Book> books = bookRepository.findByAuthorLastName("Dostoevsky");
        assertThat(books.stream().map(Book::getTitle).collect(toList()),
                equalTo(expectedBooks));
    }
}

File: src/test/kotlin/com/ddubson/repositories/BookHQLRepositoryTest.kt
package com.ddubson.repositories

import org.hamcrest.CoreMatchers.equalTo
import org.junit.Assert.assertThat
import org.junit.Test
import org.junit.runner.RunWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.test.annotation.DirtiesContext
import org.springframework.test.annotation.DirtiesContext.ClassMode
import org.springframework.test.context.junit4.SpringRunner

@RunWith(value = SpringRunner::class)
@DataJpaTest
@AutoConfigureTestDatabase
@DirtiesContext(classMode = ClassMode.BEFORE_CLASS)
class BookHQLRepositoryTest {
    @Autowired
    lateinit var bookRepository: BookHQLRepository

    @Test
    fun `findByTitle should return book with specified title`() {
        val expectedTitle = "War and Peace"
        val (_, title) = bookRepository.findByTitle(expectedTitle)
        assertThat<String>(title, equalTo<String>(expectedTitle))
    }

    @Test
    fun findByAuthorId_returnsAllBooksOfAuthor() {
        val expectedBooks = listOf("Anna Karenina", "Crime and Punishment")
        val books = bookRepository.findByAuthorId(4L)
        assertThat(books.map { it -> it.title }, equalTo(expectedBooks))
    }

    @Test
    fun findByAuthorLastName_returnsAllBooksOfAuthor() {
        val expectedBooks = listOf("Anna Karenina", "Crime and Punishment")
        val books = bookRepository.findByAuthorLastName("Dostoevsky")
        assertThat(books.map { it -> it.title }, equalTo(expectedBooks))
    }
}